xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -842,24 +868,99 @@ static String findClusterNameInRouteConfig(RouteConfiguration config, String hos
       }
     }
 
+    List<EnvoyProtoData.Route> routes = new ArrayList<>();
     // Proceed with the virtual host that has longest wildcard matched domain name with the
     // hostname in original "xds:" URI.
     // Note we would consider upstream cluster not found if the virtual host is not configured
     // correctly for gRPC, even if there exist other virtual hosts with (lower priority)
     // matching domains.
     if (targetVirtualHost != null) {
-      // The client will look only at the last route in the list (the default route),
-      // whose match field must contain a prefix field whose value is empty string
-      // and whose route field must be set.
-      List<Route> routes = targetVirtualHost.getRoutesList();
-      if (!routes.isEmpty()) {
-        Route route = routes.get(routes.size() - 1);
-        if (route.getMatch().getPrefix().isEmpty()) {
-          if (route.hasRoute()) {
-            return route.getRoute().getCluster();
-          }
+      List<Route> routesProto = targetVirtualHost.getRoutesList();
+      for (Route route : routesProto) {
+        routes.add(EnvoyProtoData.Route.fromEnvoyProtoRoute(route));
+      }
+    }
+    return routes;
+  }
+
+  /**
+   * Validates the given list of routes and returns error details if there's any error.
+   */
+  // We only validate the default route unless path matching is enabled.
```

Comment:
nit: be formal. If it is part of method specification, put it inside the head comment of this method. If it is about implementation details, put it at where the code logic is present. If it is your personal note, should just delete it.